### PR TITLE
Issue-10 parallel multi scan

### PR DESCRIPTION
**Pull request by lyndsey-ferguson**
### Checklist
- [x] I've run `bundle exec rake` from the root directory to see all new and existing tests pass
- [x] I've read the [Contribution Guidelines][contributing doc]
- [x] I've updated the documentation if necessary.

### Motivation and Context

This change adds support to run Mac and iOS tests in batches that are spread evenly across parallel test runs. Xcode 10 does support parallel test runs, but the parallelization is based on `testSuites` rather than a balanced number of tests per test run.

If you would like to participate in the beta, please modify your fastlane `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane tests:

Pluginfile:
```
gem 'fastlane-plugin-test_center', '3.8.0.parallelizing.beta.9'
```

You'll need to ensure that you include a number between 2 and 6 for `:parallel_testrun_count` to run batches of tests in parallel. For example,
```ruby
multi_scan(
  workspace: File.absolute_path('../AtomicBoy/AtomicBoy.xcworkspace'),
  scheme: 'AtomicBoy',
  try_count: 3,
  parallel_testrun_count: 4,
  fail_build: false
)
```
**Note**: if you've never used `multi_scan` before, you can read more [here](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center#multi_scan-). The **TL;DR** is that it is just like [`fastlane scan`](https://docs.fastlane.tools/actions/scan/), but with more options.

If you do discover a regression in either existing functionality, **or**, this does not parallelize for you, please create a new issue with the prefix "parallelization: " as the Issue title. If you desire new functionality, or a change (such as supporting the parallelization of invocation tests), please create a Feature Request.

#### Associated Bugs

| Issue | Fix Status |
| ----- | ------ |
|  [Parallelization Bug: Unable to find built xctestrun file](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/105) | ✅  |
| [Parallelization Bug: generated test_result bundles are deleted after completion](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/106) | ✅ |
| [Parallelization Bug: only the last completed test run block's tests are reported .test_result bundle ](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/107) | ✅   |
| [Parallelization Bug: reports for batches with spaces in the testable name are not collated properly](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/109) | ✅  |
| [Parallelization Bug: All simulators can fail with `operation never finished bootstrapping`](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/111) | ✅ |
| [Parallelization Bug: multi_scan crashes when `invocation_based_tests` is true](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/113) | ✅  |
| [Parallelization Bug: multi_scan with include_simulator_logs fails with Error 74](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/115)| ✅  |
| [Parallelization Bug: multi_scan returns false regardless of whether or not a parallel run passed or failed](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/117)| ✅  |
| [Parallelization Bug: multi_scan total_retry_count always returns nil](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/124) | 🔶 |

### Description

Here are the changes in detail:
1. Refactored the code to better reflect the patterns of serial vs. parallel test runs.
2. Refactored the code that actually runs the tests into two classes:
  1. retrying_scan which retries a test run through `scan`, but only testing failed tests.
  2. retrying_scan_helper which handles all logic of preparing for a test, providing the test options, and handling the end of a test run.
3. Created a test worker class that simply calls retrying_scan
4. Created a parallel test worker class that prepares itself for a fork of the process, runs the retrying_scan in the fork, and handles completion of the work
5. Created a test worker pool class that handles creating simulators for parallel work, creates a pool of serial or parallel workers, and runs batches of tests through the available workers.

[contributing doc]: ../docs/CONTRIBUTING.md

### CODE REVIEW

File: lib/fastlane/plugin/test_center/helper/test_collector.rb [hunk 1]
```diff
@@ -52,19 +56,30 @@ def only_testing_to_testables_tests
         tests
       end
 
+      def xctestrun_known_tests
+        config = FastlaneCore::Configuration.create(::Fastlane::Actions::TestsFromXctestrunAction.available_options, xctestrun: @xctestrun_path)
```
Comment (lyndsey-ferguson) [inline lib/fastlane/plugin/test_center/helper/test_collector.rb]: we need to pass in `invocation_based_tests: @invocation_based_tests`

File: lib/fastlane/plugin/test_center/actions/multi_scan.rb [hunk 2]
```diff
@@ -336,16 +385,22 @@ def self.example_code
           "
           UI.important(
             'example: ' \\
-            'multi_scan also works with invocation based tests.'
```
Comment (lyndsey-ferguson) [inline lib/fastlane/plugin/test_center/actions/multi_scan.rb]: we need to put this example back in

File: lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            disable_concurrent_testing: true
+          )
+          @batch_count = @test_collector.test_batches.size
+        end
+
+        def output_directory
+          @options.fetch(:output_directory, 'test_results')
+        end
+
+        def run
```
Comment (lyndsey-ferguson) [inline lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb]: add tests for this and the run of invocation tests

**Comment by szymon-kazmierczak** [comment]
This is amazing! Is this PR ready for testing? :) I'd like to run it on my projects and see if there are any issues!

**Comment by lyndsey-ferguson** [comment]
@szymon-kazmierczak you can use it for testing sure. There are somethings that I wanted to clean up before beta.

If you want, you can modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-10-parellel-multi_scan"
```

Please attach the console output as a text file to this issue (makes it easier for me to review).

**Comment by lyndsey-ferguson** [comment]
Tagging people who were interested in the fastlane issue [When Xcode 10 parallel testing is enabled, test failures look like build failures](https://github.com/fastlane/fastlane/issues/13414). @benasher44, @rockbruno, @TomCarey, @FabienLydoire, if you're interested, this may resolve the issue for you.

**Comment by yousefhamza** [comment]
Hi @lyndsey-ferguson 

When I try this build I'm always getting that [error](https://github.com/fastlane/fastlane/issues/8909), note this only happens with `invocation_based_tests`.

**Comment by lyndsey-ferguson** [comment]
@yousefhamza are you asking for log collection in your call to `multi_scan`? Can you create a new issue with a sample of your `Fastfile`?

**Comment by lyndsey-ferguson** [comment]
@yousefhamza I don't know if you actually were using the correct beta, it doesn't even work for me and I've opened up Issue 113: [Parallelization Bug: All simulators can fail with `operation never finished bootstrapping`](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/111)

**Comment by lyndsey-ferguson** [comment]
@yousefhamza I've fixed invocation tests in `3.8.0.parallelizing.beta.7`. Please try that.

**Comment by lyndsey-ferguson** [comment]
@Vkt0r @szymon-kazmierczak @ishammoho @esteluk I've fixed all known defects. I would appreciate any feedback on whether or not b7 is working for you or not. Thanks!!

**Comment by esteluk** [comment]
Thank you so much, will try to give it a go tomorrow :) I got a little lost trying to work out the expected behaviour for collating the reports at the end when I had an attempt last week.

File: lib/fastlane/plugin/test_center/helper/multi_scan_manager/simulator_helper.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+              cloned_simulator.rename("#{new_first_name} #{new_last_name}")
+
+              cloned_simulators.last << cloned_simulator
+            end
+          end
+          cloned_simulators
+        end
+
+        def delete_multi_scan_cloned_simulators
+          FastlaneCore::DeviceManager.simulators('iOS').each do |simulator|
```
Comment (lyndsey-ferguson) [inline lib/fastlane/plugin/test_center/helper/multi_scan_manager/simulator_helper.rb]: verbose message about deletion

File: spec/multi_scan_manager/retrying_scan_helper_spec.rb [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+          test_operation_failure: 'Launch session expired before checking in',
+          batch: 1,
+          try_count: 1,
+          report_filepath: nil
+        )
+      end
+    end
+  end
+end
+
```
Comment (lyndsey-ferguson) [inline spec/multi_scan_manager/retrying_scan_helper_spec.rb]: remove blank space

File: spec/multi_scan_spec.rb [hunk 6]
```diff
@@ -1,180 +1,322 @@
...
+        raise_error(FastlaneCore::Interface::FastlaneError) do |error|
+          expect(error.message).to match(
+            "Error: Can't use 'invocation_based_tests' and 'batch_count' options in one run, "\
+            "because the number of tests is unkown."
+          )
+        end
+      )
+    end
   end
+
```
Comment (lyndsey-ferguson) [inline spec/multi_scan_spec.rb]: remove space

File: spec/quit_core_simulator_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        result = QuitCoreSimulatorServiceAction.run(config)
+        expect(result).to eq(['launchctl remove'])
+      end
+
+      it 'crashes after 11 attempts' do
+        allow(Fastlane::Actions).to receive(:sh)
+          .with(/launchctl list/, anything)
+          .and_return('running')
+
+
```
Comment (lyndsey-ferguson) [inline spec/quit_core_simulator_service_spec.rb]: remove extra space

**Comment by lyndsey-ferguson** [comment]
I've published a new version `3.8.0.parallelizing.beta.8` that has console log messages cleaned up and extra helpful logging messages accessible when the `--verbose` flag is sent.

**Comment by lyndsey-ferguson** [comment]
> @Vkt0r @szymon-kazmierczak @ishammoho @esteluk I've fixed all known defects. I would appreciate any feedback on whether or not b7 is working for you or not. Thanks!!

Any feedback? If you don't have time to write, _but have_ tried it, you can respond with a thumbs up here 😄

**Comment by Vkt0r** [comment]
@lyndsey-ferguson I missed that message before, sorry about that! I would try it during the rest of the weekend and I would give you feedback about it.

**Comment by yousefhamza** [comment]
Hi, @lyndsey-ferguson I just got back to work and I got to try your new build.

My setup is we have 3 test suits: 2 unit tests (1 XCTest, 1 Invocation based tests) and 1 UI tests

I'm using the `parallel_count` only on the UITest test suit and the issues I'm getting with the beta build is:
Unit tests (XCTest) ✅
====================================================
Unit tests (Invocation based tests) 🔴
Fails with exit code 74 when collecting logs which I set to `false` in my fastfile:
```
result = multi_scan(
      workspace: "Instabug/Instabug.xcworkspace",
      try_count: 3,
      fail_build: true,
      include_simulator_logs: false,
      invocation_based_tests: true,
      device: 'iPhone 8 (12.2)',
      scheme: 'AllUnitTests'
    )
```
but the option get bassed to `scan` as `true`
====================================================
The UITests _(which uses the new parallezation feature)_  ⚠️
Works fine. But when a test fails in the middle and pass later when retried `multi_scan` report that tests have failed.
```
multi_scan(
      workspace: "Instabug/Instabug.xcworkspace",
      try_count: 3,
      fail_build: true,
      device: 'iPhone 8 (12.2)',
      parallel_testrun_count: 2,
      scheme: 'InstabugDemoUITests',
      include_simulator_logs: false
    )
```

**Comment by lyndsey-ferguson** [comment]
@Vkt0r 🙏 Please and thank you!

**Comment by lyndsey-ferguson** [comment]
@yousefhamza thanks for testing. Can you confirm what I understand for the first issue you report?

Without the `:parallel_testrun_count` option, you provide `multi_scan` the `:fail_buid` option as `false`, but you can see it being passed to `scan` in the summary as `true`? So, that would be one bug.

**Comment by lyndsey-ferguson** [comment]
@yousefhamza for the second issue, can you confirm my understanding?

**With** the `:parallel_testrun_count` option set to 2, you see that a test run fail, then retry and the previously failing tests pass, and then `multi_scan` reports this as a failure?

**Comment by yousefhamza** [comment]
@lyndsey-ferguson first issue, Yes but `:include_simulator_logs` instead of `:fail_build`
Second issue Yes.

**Comment by lyndsey-ferguson** [comment]
@yousefhamza I've created [Parallelization Bug: multi_scan with include_simulator_logs fails with Error 74](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/115) for issue one. Can you confirm my understanding of the second issue that you mentioned?

**Comment by Vkt0r** [comment]
@lyndsey-ferguson I a preliminary test I conducted in a personal repo during the weekend only for UI Tests I'm experiencing the same result of @yousefhamza. I tested with 2-4 parallel simulators at least 5 times and I always have some tests failing in the middle that later passed and the build fails at the end.

I would be conducting a bigger test today with a bigger UI Test suite to see and compare results. I'll let you know

**Comment by lyndsey-ferguson** [comment]
Thanks @Vkt0r. So the second issue is that even after a retry, the final result is "fail" 😒

**Comment by Vkt0r** [comment]
Yes, @lyndsey-ferguson is failing even when the retry is successful. I'll be throwing more results today.

**Comment by Vkt0r** [comment]
@lyndsey-ferguson Another issue I've noticed is still happening is #107 when I run it with 4 simulators in a big UI Tests suite. I going to set `collate_reports: false` to see if there are being generated correctly independently.

**Comment by lyndsey-ferguson** [comment]
@Vkt0r make sure that you're using b8, and if it is still happening, we can re-open that Issue if you include verbose logs and an example Fastfile showing how you are calling it.

**Comment by Vkt0r** [comment]
@lyndsey-ferguson I'm using `b8`, pretty sure of that `fastlane-plugin-test_center (= 3.8.0.parallelizing.beta.8)` in my `Gemfile.lock`. Unfortanely, I don't think I could include the verbose logs for privacy matters, if there is any other way I can help you let me know. Below is how the plugin is called:

```ruby
desc "Run test in the app"
private_lane :tests do |options|

  # clear the derived data from Xcode
  clear_derived_data

  # run tests
  multi_scan(
    try_count: 3,
    device: options[:device],
    fail_build: true,
    scheme: options[:scheme],
    configuration: 'Debug',
    result_bundle: true,
    quit_simulators: false,
    workspace: workspace,
    buildlog_path: "./fastlane/test_output",
    parallel_testrun_count: 4,
    collate_reports: false
  )
end
```

In the first test with `collate_reports: false` it generated successfuly the `resultBundle` for each batch and for each retry in each batch, nevertheless when I collated the `resultBundle` generated in the batch 1 using XCHTMLReport I noticed inconsistencies between the output and the number of failed tests reported so I want to verify it again in another run. I'll keep you posted.

**Comment by lyndsey-ferguson** [comment]
@Vkt0r 

> nevertheless when I collated the resultBundle generated in the batch 1 using XCHTMLReport I noticed inconsistencies between the output and the number of failed tests reported so I want to verify it again in another run.

Can you clarify, I am uncertain if you are saying that you are collating the result bundles with `XCHTMLReport` or if you are using the `collate_test_result_bundles` action to collate them. 

If there is a way you can search and replace the sensitive data out and provide the logs, that may help me. 

And to make it explicit, the bug that you are encountering is that only the _last_ `test_result` bundle is ever being reported? That there are no intermediate `test_result` bundles? If you include a `testrun_completed_block` and print out a Dir.glob("#{outputdirectory}/**/*.test_result") it only would print out one `test_result` bundle?

```ruby
outputdirectory = File.absolute_path('test_output')

test_run_block = lambda do |testrun_info|
   test_results = Dir.glob("#{outputdirectory}/**/*.test_result")
   puts test_results
end

multi_scan(
  try_count: 3,
    device: options[:device],
    fail_build: true,
    scheme: options[:scheme],
    configuration: 'Debug',
    result_bundle: true,
    quit_simulators: false,
    workspace: workspace,
    buildlog_path: "./fastlane/test_output",
    parallel_testrun_count: 4,
    testrun_completed_block: test_run_block,
    output_directory: outputdirectory
)
```

**Comment by ishammoho** [comment]
sorry @lyndsey-ferguson 
I was away, I'll try this today and let you know how it goes. Thanks :)

**Comment by Vkt0r** [comment]
@lyndsey-ferguson 

>Can you clarify, I am uncertain if you are saying that you are collating the result bundles with XCHTMLReport or if you are using the collate_test_result_bundles action to collate them.

Sure, I was referring to XCHTMLReport, not to the `collate_test_result_bundles` action. But in fact it wasn't an issue of XCHTMLReport neither as I was collating different `resultBundle` but of the same batch and in one `resultBundle` the test failed but in another later passed but XCHTMLReport is not configured to deal with the same test failing and passing in several `resultBundlePath`.

>And to make it explicit, the bug that you are encountering is that only the last test_result bundle is ever being reported? 

Yes. The case I mentioned above was with the `collate_reports: false`. When I executed without `collate_reports` there was only one `resultBundlePath` result of being collated and it was only of the last worker.

>If you include a testrun_completed_block and print out a Dir.glob("#{outputdirectory}/**/*.test_result") it only would print out one test_result bundle?

I'll execute it with the `testrun_completed_block ` now to see the result. I'll keep you posted.

>If there is a way you can search and replace the sensitive data out and provide the logs, that may help me.

I'll try to create a sample project as soon as I can to try to recreate the possible scenarios.

**Comment by lyndsey-ferguson** [comment]
@Vkt0r @yousefhamza  I've created [Parallelization Bug: multi_scan returns false regardless of whether or not a parallel run passed or failed](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/117)

**Comment by lyndsey-ferguson** [comment]
# Everyone
All **known** issues with the parallelization have been fixed. Could you please try it and confirm if everything still works for you?

If it does, I will be releasing this as the new official version. The faster I get confirmations, the sooner this will be released.

Once again, thank you so much for your participation and activity on this. Your help has been essential to putting together a solid release.

**Comment by lyndsey-ferguson** [comment]
@Vkt0r @szymon-kazmierczak @ishammoho @esteluk @ishammoho ^^ 👀

**Comment by lyndsey-ferguson** [comment]
Just a clarification, I'm not looking for thorough testing as you've already done that. Just **one** more run of what you've already performed with the `parallel_testrun_count` option > 1.

**Comment by ishammoho** [comment]
Have tested with 4 parallel runs. It's working fine. Except for HTML report issues. Thanks

**Comment by lyndsey-ferguson** [comment]
@ishammoho that sounds almost good, what are the HTML report issues? Are you referring to how the `test_result` bundles are not collated in a way that the external Ruby gem, `XCHTMLReport`, can process? If so, please create a ticket, otherwise, please elaborate.

**Comment by ishammoho** [comment]
I'm struggling to generate any sort of html report. 
my lane looks like this
```
result = multi_scan(
        scheme: 'Automation',
        try_count: 3,
        parallel_testrun_count: 4,
        fail_build: false,
        result_bundle: true,
        output_files: 'report.html',
        output_types: 'html',
        testrun_completed_block: test_run_block
      )
```

however the html report generates from this is blank. But not an issue for us at the moment. We are using the junit reports. But just reporting my findings.

**Comment by lyndsey-ferguson** [comment]
@ishammoho hm. Mine are not empty, but they are not pretty. The collation seems to remove the tests out of the nice format that they were in.
Screen Shot 2019-08-06 at 8 03 14 PM

**Comment by lyndsey-ferguson** [comment]
Okay, I've confirmed. This is an issue with the `v3.7.0` as well.

**Comment by ishammoho** [comment]
@lyndsey-ferguson Thanks. Can you please share your lane pls. Just to see whether I'm doing anything wrong.

**Comment by lyndsey-ferguson** [comment]
@ishammoho let's continue the HTML conversation [in this Issue](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/issues/126)

**Pull request merged**